### PR TITLE
Pipes - Eva - MediaRanker

### PR DESCRIPTION
**Pull request by Esawyer25**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Works#top_ten_albums, which makes an array of the top ten albums.  Initially I had a method that made an array of the top ten of whatever type of thing you passed into it, but I got rid of it when I was trying to troubleshoot deployment... turned out that was not the problem with deployment though.
Describe how you approached testing that model method. What edge cases did you come up with? | I did not test this model successfully.  I wanted to test how it behaved with more than ten works, less than ten but more than 0 works, and 0 works.  However, I forgot that the work_ids would be randomly set in to large numbers, so when I tried to hardcode some votes for my works I was not able to get the vote.work_ids to match the work.ids.  Thus, all I ended up testing was if the top_ten methods made arrays.  They make arrays. 
Describe an edge case test you wrote for a controller | Does the controller action still work if I delete all the instances of that object in the DB.
What are `session` and `flash`? What is the difference between them? | They are both special hashes that can keep track of the program over more than one request cycle.  Flash resets after two cycles, session does not reset until you reset it.
Describe a controller filter you wrote. | In my WorkController I had a before_action block that ran a private method find_by_param_id before several methods. I 
What was one thing that you gained more clarity on through this assignment? |  Very many things, including yml files, controller testing, session, flash, foundation.  Also this was the first project that really felt so big that it could not fit in my head all at the same time, which was difficult to work with.
What is the Heroku URL of your deployed application | https://vast-wildwood-85785.herokuapp.com/ , but it doesn't work.  An exert from the logs is below.  I deleted all my data and the project runs fine on my server, so I don't think the problem is b/c there is nothing in the database.  I think the problem has something to do with this line "ActiveRecord::StatementInvalid (PG::UndefinedTable: ERROR:  relation "works" does not exist", and I think that means there is a problem in the way my relations are set up in the empty database so that initially "work.vote" does not work when i call it in "app/models/work.rb:24:in `sort_by'"  , but I couldn't figure out how to fix it. 

I, [2017-10-16T03:01:44.705375 #4]  INFO -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39] Processing by MainController#index as HTML
2017-10-16T03:01:44.717382+00:00 app[web.1]: D, [2017-10-16T03:01:44.717224 #4] DEBUG -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39]   Work Load (2.0ms)  SELECT "works".* FROM "works"
2017-10-16T03:01:44.717678+00:00 app[web.1]: I, [2017-10-16T03:01:44.717602 #4]  INFO -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39] Completed 500 Internal Server Error in 12ms (ActiveRecord: 2.0ms)
2017-10-16T03:01:44.718388+00:00 app[web.1]: F, [2017-10-16T03:01:44.718324 #4] FATAL -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39]   
2017-10-16T03:01:44.718477+00:00 app[web.1]: F, [2017-10-16T03:01:44.718388 #4] FATAL -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39] ActiveRecord::StatementInvalid (PG::UndefinedTable: ERROR:  relation "works" does not exist
2017-10-16T03:01:44.718479+00:00 app[web.1]: LINE 1: SELECT "works".* FROM "works"
2017-10-16T03:01:44.718481+00:00 app[web.1]:                               ^
2017-10-16T03:01:44.718482+00:00 app[web.1]: : SELECT "works".* FROM "works"):
2017-10-16T03:01:44.718534+00:00 app[web.1]: F, [2017-10-16T03:01:44.718470 #4] FATAL -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39]   
2017-10-16T03:01:44.718607+00:00 app[web.1]: F, [2017-10-16T03:01:44.718544 #4] FATAL -- : [b7d1bd7a-70fd-462b-b393-b5e42c79dc39] app/models/work.rb:24:in `sort_by'
2017-10-16T03:01:44.718608+00:00 app[web.1]: [b7d1bd7a-70fd-462b-b393-b5e42c79dc39] app/models/work.rb:24:in `top_ten_albums'
2017-10-16T03:01:44.718608+00:00 app[web.1]: [b7d1bd7a-70fd-462b-b393-b5e42c79dc39] app/controllers/main_controller.rb:5:in `index'

Do you have any recommendations on how we could improve this project for the next cohort? | This was the first project I struggled to finish.  If many others also have that problem, and you are not aiming to have us experience this struggle, then it was too hard/ there was not enough time.  Otherwise, it was probably a fine experience for me to be challenged so.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  get "main/index"
+
+  resources :works do
+     resources :votes, only: [:create]
+   #post '/votes', to: 'work#vote'
+  #   # resources :votes, only: [:new]
+   end
+   
+  resources :users, only: [:index, :show, :login, :logout]
+
```
Comment (droberts-sea) [inline config/routes.rb:14]: I don't think that the `:login, :logout` in the `only` argument will do anything.

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+<html>
+<head>
+  <title>MediaRanker</title>
+  <%= csrf_meta_tags %>
+  <%= stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload' %>
+  <%= javascript_include_tag 'application', 'data-turbolinks-track': 'reload' %>
+</head>
+
+<header class = "page-header row">
+  <h2>
```
Comment (droberts-sea) [inline app/views/layouts/application.html.erb:11]: All of this should be inside the `<body>` tag.

File: app/views/main/_top_ten.html.erb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+
+<ul>
+  <% collection_of_media.each do |work|%>
```
Comment (droberts-sea) [inline app/views/main/_top_ten.html.erb:3]: Good use of a partial to DRY this up.

File: test/controllers/votes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+require "test_helper"
+
+describe VotesController do
+  def login
+    valid_params = {}
+    valid_params[:user] = {name: "koko the gorilla"}
```
Comment (droberts-sea) [inline test/controllers/votes_controller_test.rb:6]: Good use of a helper method here

File: test/controllers/works_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+      must_respond_with :bad_request
+
+      work.reload
+      work.title.wont_equal bogus_work_update[:work][:title]
+    end
+  end
+
+  describe "destroy" do
+    it "returns success and destroys work when given a valid work ID" do
+      work_id = Work.first.id
```
Comment (droberts-sea) [inline test/controllers/works_controller_test.rb:140]: What about a work that has votes? Can you destroy it? What happens to the votes?

File: test/fixtures/votes.yml [hunk 6]
```diff
@@ -0,0 +1,185 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  user_id: 1
+  work_id: 1
```
Comment (droberts-sea) [inline test/fixtures/votes.yml:5]: Instead of specifying relations by ID, you should use the names you assigned in the other yaml files. So this vote would look like:

```yaml
one:
  user: one
  work: one
```

File: test/fixtures/works.yml [hunk 7]
```diff
@@ -0,0 +1,232 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+one:
+  category: album
+  title: album_1
+  creator: album_creator_1
```
Comment (droberts-sea) [inline test/fixtures/works.yml:6]: You can stick ERB in your yaml files. Something like:

```
<% categories = ["album", "movie", "book"] %>
<% 10.times do |i| %>
work_<%= i %>:
  category: <%= category[i % 3] %>
  title: work_<%= i %>
  creator: some artist
# ...
```

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      Work.destroy_all
+      top_books = Work.top_ten_books
+      top_books.must_be_kind_of Array
+      top_books.length.must_equal 0
+    end
+  end
+
+  describe "self.top_work" do
+    it "returns the book with the most votes when their are works" do
+      top = Work.top_work
```
Comment (droberts-sea) [inline test/models/work_test.rb:107]: What if there are no works, or no votes? What if there's a tie?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes - For deployment, did you do `heroku run rails db:migrate` to make sure your database is set up?
**General** |
Rails fundamentals (RESTful routing, use of named paths) | yes
Semantic HTML | yes
Errors are reported to the user | yes
Business logic lives in the models | yes - good work
Models are thoroughly tested, including relations, validations and any custom logic | yes - There are a few places where this could be expanded, some of which you've talked about in your PR and some I've pointed out inline. Be vigilant about finding edge cases.
Controllers are thoroughly tested | yes - great work
**Wave 1 - Media** |
Splash page shows the three media categories| yes
Basic CRUD operations on media are present and functional | yes
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
Individual user pages and the user list are present | yes
A user cannot vote for the same media more than once | yes
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
Media pages contain lists of voting users | yes
**Wave 3 - Styling** |
Foundation is used appropriately | yes
Look and feel is similar to the original | yes
**Overall** | Great job overall! This is a complex project with many moving parts, and I'm quite happy with this submission. Keep up the hard work.
